transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java

```diff
@@ -0,0 +1,266 @@
+            @Override
+            protected void initChannel(Channel ch) throws Exception {
+                assert ch.eventLoop().inEventLoop();
+                handler.channelCreated(ch);
+            }
+        });
+    }
+
+    @Override
+    public final Future<Channel> acquire() {
+        return acquire(bootstrap.group().next().<Channel>newPromise());
+    }
+
+    @Override
+    public Future<Channel> acquire(final Promise<Channel> promise) {
+        checkNotNull(promise, "promise");
+        try {
+            final Channel ch = pollChannel();
+            if (ch == null) {
+                // No Channel left in the pool bootstrap a new Channel
+                Bootstrap bs = bootstrap.clone();
+                bs.attr(POOL_KEY, this);
+                ChannelFuture f = connectChannel(bs);
+                if (f.isDone()) {
+                    notifyConnect(f, promise);
+                } else {
+                    f.addListener(new ChannelFutureListener() {
+                        @Override
+                        public void operationComplete(ChannelFuture future) throws Exception {
+                            notifyConnect(future, promise);
+                        }
+                    });
+                }
+                return promise;
+            }
+            EventLoop loop = ch.eventLoop();
+            if (loop.inEventLoop()) {
+                doHealthCheck(ch, promise);
+            } else {
+                loop.execute(new OneTimeTask() {
+                    @Override
+                    public void run() {
+                        doHealthCheck(ch, promise);
+                    }
+                });
+            }
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private static void notifyConnect(ChannelFuture future, Promise<Channel> promise) {
+        if (future.isSuccess()) {
+            promise.setSuccess(future.channel());
+        } else {
+            promise.setFailure(future.cause());
+        }
+    }
+
+    private void doHealthCheck(final Channel ch, final Promise<Channel> promise) {
+        assert ch.eventLoop().inEventLoop();
+
+        Future<Boolean> f = healthCheck.isHealthy(ch);
+        if (f.isDone()) {
+            notifyHealthCheck(f, ch, promise);
+        } else {
+            f.addListener(new FutureListener<Boolean>() {
+                @Override
+                public void operationComplete(Future<Boolean> future) throws Exception {
+                    notifyHealthCheck(future, ch, promise);
+                }
+            });
+        }
+    }
+
+    private void notifyHealthCheck(Future<Boolean> future, Channel ch, Promise<Channel> promise) {
+        assert ch.eventLoop().inEventLoop();
+
+        if (future.isSuccess()) {
+            if (future.getNow() == Boolean.TRUE) {
+                try {
+                    ch.attr(POOL_KEY).set(this);
```

Comment:
it is as the bs.attr(...) only is used for new channels while this is for already pooled channel.s